### PR TITLE
API reference generated from Open API spec file

### PR DESCRIPTION
**Pull request by lewisnyman**
This branch allows you to define a path to an Open API file in tech-docs.yml: `api_path: source/pets.yml`. This can be a relative file path or a URI to a raw file elsewhere.

To output the generated content in the page of your choice, write `api>` in any markdown parsed file (.md, .html.md, .html.md.erb). You can also specify to just print a specific path: `api> /pets`

See the example project on how to set this up.

### CODE REVIEW

**Comment by lewisnyman** [comment]
I've got an experimental branch attempting to add in some unit tests but it requires a big restructure as it seems to be very difficult to unit test a Middleman extension. I'm not even sure which functions are worth unit testing that isn't tested in the integration test as this extension doesn't have complex logic.

**Comment by MatMoore** [comment]
I haven't looked at the code yet, but here's some initial feedback from looking at the pet store example.

I'll take a deeper look later today or tomorrow and check it against some different specs. With the previous solution I had a lot of trouble documenting responses with nested objects and arrays, so I'm keen to try out that use case.

## Things that don't work well
On the "API /Pets" page where you embed the endpoint docs, the page doesn't include the schema section, which means that the schema links in the responses table don't work (the "Swagger Petstore v1.0.0" pages render these sections so the links work there).

I think we also need a page in the manual that explains the `api > /pets` syntax and when you might want to use it.

In the schema section, the headings are a bit confusing. There is an empty section rendered for "Pets" and the Error and Pet schemas have a subheading "properties" which could be removed.

## Customisability
It seems like at the moment you have to choose whether you want everything autogenerated from the spec, or you have to create a markdown file for every endpoint, which is fine, but the ideal for me is probably somewhere in the middle.

For example, with the [registers tech docs](https://docs.registers.service.gov.uk/api_reference/#api-reference) we moved "Rate limits" and "HTTP status codes" into the API reference section, and I'm not sure we'd still be able to do this if we were generating all the endpoint documentation from the spec.

When I tried out [dapperdox](http://dapperdox.io/) a while back it was easier to customise, because the page is divided into multiple components and you can choose which bits you want to override. 

Personally I would prefer if there was a separate page for every resource by default, rather than having it all on one page, because it feels quite information-dense. For a real API, the property tables can get very long, and I find it a bit annoying when if I scroll a bit too far I see information for some unrelated endpoint. This could be addressed in other ways though - [Stripe do the single page thing very well](https://stripe.com/docs/api#customer_object), but they also have a lot of vertical space between stuff, which makes it easier to read.

**Comment by MatMoore** [comment]
@lewisnyman I don't know how viable this is, but I was thinking that if I was using `api>`, it would be great if you could just dereference different parts of the spec like

- `api>/pets>get>parameters`
- `api>/pets>get>responses>200`
- `api>schemas>Pet`

and be able to render them inline. But I would definitely consider this a 'nice to have' as I don't know whether I'd actually use the `api>` method at all.

File: lib/govuk_tech_docs/api_reference/api_reference.rb [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        else
+          return text
+        end
+      else
+        text
+      end
+    end
+
+    def api_path_render(text)
+      if text == 'api&gt;'
```
[the remark this answers is not in the document]
Comment (mikebell) [reply lib/govuk_tech_docs/api_reference/api_reference.rb]: Yes it does from my testing.

File: lib/govuk_tech_docs/api_reference/api_reference.rb [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    rescue URI::BadURIError
+      false
+    rescue URI::InvalidURIError
+      false
+    end
+
+    def api(text)
+      if @api_parser == true
+
+        map = {
```
[the remark this answers is not in the document]
Comment (mikebell) [reply lib/govuk_tech_docs/api_reference/api_reference.rb]: Renamed to keywords in a new commit.

File: lib/govuk_tech_docs/api_reference/api_reference.rb [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+      @config = @app.config[:tech_docs]
+
+      # If no api path then just return.
+      if @config['api_path'].to_s.empty?
+        # @TODO Throw a middleman error?
+        return
+      end
+
+      # Is the api_path a url or path?
+      if uri?@config['api_path']
```
[the remark this answers is not in the document]
Comment (mikebell) [reply lib/govuk_tech_docs/api_reference/api_reference.rb]: Fixed in new commit

File: lib/govuk_tech_docs/api_reference/api_reference.rb [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+        md = text.match(/^<p>(#{regexp})/)
+        if md
+          key = md.captures[0]
+          type = map[key]
+
+          text.gsub!(/#{ Regexp.escape(key) }\s+?/, '')
+
+          # Strip paragraph tags from text
+          text = text.gsub(/<\/?[^>]*>/, '')
```
[the remark this answers is not in the document]
Comment (mikebell) [reply lib/govuk_tech_docs/api_reference/api_reference.rb]: Not at the moment, it only works if that's the only thing on the line. Is there a use case where you'd want to include this in a paragraph?

File: lib/govuk_tech_docs/api_reference/api_reference.rb [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    def initialize(app, options_hash = {}, &block)
+      super
+
+      @app = app
+      @config = @app.config[:tech_docs]
+
+      # If no api path then just return.
+      if @config['api_path'].to_s.empty?
+        # @TODO Throw a middleman error?
```
[the remark this answers is not in the document]
Comment (mikebell) [reply lib/govuk_tech_docs/api_reference/api_reference.rb]: Fixed in latest commit.

File: lib/govuk_tech_docs/api_reference/api_reference.rb [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+      # Is the api_path a url or path?
+      if uri?@config['api_path']
+        @api_parser = true
+        @document = Openapi3Parser.load_url(@config['api_path'])
+      elsif File.exist?(@config['api_path'])
+        # Load api file and set existence flag.
+        @api_parser = true
+        @document = Openapi3Parser.load_file(@config['api_path'])
+      else
+        # @TODO Throw a middleman error?
```
[the remark this answers is not in the document]
Comment (mikebell) [reply lib/govuk_tech_docs/api_reference/api_reference.rb]: Fixed in latest commit

File: lib/govuk_tech_docs/api_reference/api_reference.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (MatMoore) [reply lib/govuk_tech_docs/api_reference/api_reference.rb]: I don't think so - requiring it to be in its own paragraph makes sense I think.

**Comment by jonathanglassman** [comment]
@MatMoore travis check failed :(

**Pull request merged**